lib/jwe.js

```diff
@@ -0,0 +1,511 @@
+      var str = (n*j+i).toString(16);
+      if (str.length %2 !== 0)
+        str = '0' + str;
+      var t = createBuffer(str, 'hex');
+
+      // B = AES-1(K, (A^t) | R[i])
+      var aes = crypto.createDecipheriv(algorithm, K, '');
+      aes.setAutoPadding(false);
+      var B = aes.update(Buffer.concat([xor(A, t), R[i]]), null, 'hex');
+      B += aes.final('hex');
+      B = createBuffer(B, 'hex');
+
+      // A = MSB(64, B)
+      A = B.slice(0, 8);
+
+      // R[i] = LSB(64, B)
+      R[i] = B.slice(B.length-8);
+    }
+  }
+
+  /****************************************************************************
+   * 3) Output results.
+   *    If A is an appropriate initial value
+   *    Then
+   *      For i = 1 to n
+   *        P[i] = R[i]
+   *    Else
+   *      Return an error
+   ***************************************************************************/
+  
+  // check A
+  if (A.toString('hex').toUpperCase() === 'A6A6A6A6A6A6A6A6') {
+    var result = R[1];
+    for (var i = 2; i <= n; i++)
+      result = Buffer.concat([result, R[i]]);
+    
+    return result;
+  } else {
+    throw new Error('aes decryption failed: invalid key');
+  }
+};
+
+/******************************************************************************
+ * AES-CBC-HMAC-SHA2 decryption
+ *****************************************************************************/
+
+/**
+ * decrypt ciperText using aes-cbc-hmac-sha2 algorithm
+ *
+ * @param{String}  algorithm: algorithm
+ * @param{Buffer}  key: encryption key
+ * @param{Buffer}  cipherText: encrypted content
+ * @param{Buffer}  iv: intialization vector
+ * @param{Buffer}  aad: additional authentication data
+ * @param{Buffer}  authTag: authentication tag
+ * @return{Buffer}  decrypted content
+ */
+var decrypt_AES_CBC_HMAC_SHA2 = (algorithm, key, cipherText, iv, aad, authTag) => {
+  // algorithm information
+  // note ENC_KEY_LEN = MAC_KEY_LEN = T_LEN = len
+  var algInfo = {
+    'aes-128-cbc-hmac-sha-256': {
+        'cipher_algo': 'aes-128-cbc', 
+        'hash_algo': 'sha256', 
+        'len': 16
+      },
+    'aes-192-cbc-hmac-sha-384': {
+        'cipher_algo': 'aes-192-cbc', 
+        'hash_algo': 'sha384', 
+        'len': 24
+      },
+    'aes-256-cbc-hmac-sha-512': {
+        'cipher_algo': 'aes-256-cbc', 
+        'hash_algo': 'sha512', 
+        'len': 32
+      }
+  };
+
+  // 1. Verify input
+
+  // check if we have a supported algorithm
+  if (!algorithm)
+    throw new Error('In decrypt_AES_CBC_HMAC_SHA2: algorithm is not provided');
+  if (!algInfo[algorithm])
+    throw new Error('In decrypt_AES_CBC_HMAC_SHA2: unsupported algorithm: ' + algorithm);
+
+  var algo = algInfo[algorithm];
+
+  // check the size of key
+  if (!key)
+    throw new Error('In decrypt_AES_CBC_HMAC_SHA2: key is not provided');
+  if (!(key instanceof Buffer) || key.length !== algInfo[algorithm].len * 2)
+    throw new Error('In decrypt_AES_CBC_HMAC_SHA2: key must be a buffer of size ' + algo.len * 2);
```

Comment:
can you add the size of the key to the error message? Will help users if they get the params mixed up.